src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,952 @@
+
+
+
+  let assign_texpr t var texp = timing_wrap "assign_texpr" (assign_texpr t var) texp
+
+  (* no_ov -> no overflow
+     if it's true then there is no overflow
+      -> Convert.texpr1_expr_of_cil_exp handles overflow *)
+  let assign_exp (t: VarManagement.t) var exp (no_ov: bool Lazy.t) =
+    let t = if not @@ Environment.mem_var t.env var then add_vars t [var] else t in
+    match Convert.texpr1_expr_of_cil_exp t t.env exp (Lazy.force no_ov) with
+    | exp -> assign_texpr t var exp
+    | exception Convert.Unsupported_CilExp _ ->
+      if is_bot_env t then t else forget_vars t [var]
+
+  let assign_exp t var exp no_ov =
+    let res = assign_exp t var exp no_ov in
+    if M.tracing then M.tracel "ops" "assign_exp t:\n %s \n var: %s \n exp: %a\n no_ov: %b -> \n %s\n"
+        (show t) (Var.to_string var) d_exp exp (Lazy.force no_ov) (show res) ;
+    res
+  let assign_var (t: VarManagement.t) v v' =
+    let t = add_vars t [v; v'] in
+    let texpr1 = Texpr1.of_expr (t.env) (Var v') in
+    assign_texpr t v (Apron.Texpr1.to_expr texpr1)
+
+  let assign_var t v v' =
+    let res = assign_var t v v' in
+    if M.tracing then M.tracel "ops" "assign_var t:\n %s \n v: %s \n v': %s\n -> %s\n" (show t) (Var.to_string v) (Var.to_string v') (show res) ;
+    res
+  (* This functionality is not common to C and is used for assignments of the form: x = y, y=x; which is not legitimate C grammar
+     x and y should be assigned to the value of x and y before the assignment respectively.
+     ==> x = y_old , y = x_old;
+     Therefore first apply the assignments to temporary variables x' and y' to keep the old dependencies of x and y 
+     and in a second round assign x' to x and y' to y
+  *)
+  let assign_var_parallel t vv's = 
+    let assigned_vars = List.map (function (v, _) -> v) vv's in
+    let t = add_vars t assigned_vars in
+    let primed_vars = List.init (List.length assigned_vars) (fun i -> Var.of_string (Int.to_string i  ^"'")) in (* TODO: we use primed vars in analysis, conflict? *)
+    let t_primed = add_vars t primed_vars in
+    let multi_t = List.fold_left2 (fun t' v_prime (_,v') -> assign_var t' v_prime v') t_primed primed_vars vv's in
+    match multi_t.d with
+    | Some arr when not @@ is_top multi_t -> 
+      let switched_arr = List.fold_left2 (fun multi_t assigned_var primed_var-> assign_var multi_t assigned_var primed_var) multi_t assigned_vars primed_vars in
+      let res = drop_vars switched_arr primed_vars true in
+      let x = Option.get res.d in
+      {d = Some x; env = res.env} 
+    | _ -> t
+
+  let assign_var_parallel t vv's =
+    let res = assign_var_parallel t vv's in
+    if M.tracing then M.tracel "ops" "assign_var parallel: %s -> %s \n" (show t) (show res);
+    res
+
+  let assign_var_parallel t vv's = timing_wrap "var_parallel" (assign_var_parallel t) vv's
+
+  let assign_var_parallel_with t vv's =
+    let t' = assign_var_parallel t vv's in
+    t.d <- t'.d;
+    t.env <- t'.env
```

Comment:
One of the main ideas of having the `_with` methods that modify the input data is to be more efficient, by not having to allocate more additional data structures than needed for in-place modification. Would this be possible to do here? 
One would then implement the `_with` version of the function, and delegate to it in the version of the function that makes a copy first.

Reply:
The environment changes can not be done in place as these are using the library functions from Apron.Environment. Assign_var_parallel is also a function which is likely never called by goblint as such parallel assignments are not supported by C. As a result, changing this for this specific function would probably mean unnecessary effort.

For other functions this is likely possible to some degree but honestly, I don't know at which library functions of Ocaml in general and Apron a new object or copy is generated, therefore such design changes might change the whole coding style to a less functional style and would need quite the effort.

Reply:
> The environment changes can not be done in place as these are using the library functions from Apron.Environment.

I see!


> Assign_var_parallel is also a function which is likely never called by goblint

`assign_var_parallel_with` is used whenever a function is entered (`body`), in `unlock`, at `sync` edges, and when entering multi-threaded mode.

Maybe we can discuss these `_with` versions of functions once the rest is cleaned up.

Reply:
This sounds good :)

Reply:
Added comment about this unresolved TODO.